## TDD red output

Failing-test commit: `50da42167 test(release-video): cover label-only visual cues`

Command:

```bash
conda run -n pdd pytest -q tests/test_release_video.py::test_release_video_normalizes_label_only_visual_blocks tests/test_release_video.py::test_release_video_does_not_swallow_unsafe_label_only_visual_blocks
```

Output excerpt:

```text
FF                                                                       [100%]
=================================== FAILURES ===================================
____________ test_release_video_normalizes_label_only_visual_blocks ____________

>       assert "\nVISUAL:\n" not in artifacts["script"]
E       AssertionError: assert '\nVISUAL:\n' not in '# PDD v1.1....nge entry.\n'
E         '\nVISUAL:\n' is contained here:
E           nything.
E           
E           VISUAL:
E           
E           show the release context, normalized script, and pds_run.json side by side with callout labels for raw output, final script, and validation state....

tests/test_release_video.py:424: AssertionError
_____ test_release_video_does_not_swallow_unsafe_label_only_visual_blocks ______

>       assert "\nVISUAL: show the validation JSON beside the final PDS create command" in artifacts["script"]
E       AssertionError: assert '\nVISUAL: show the validation JSON beside the final PDS create command' in '# PDD v1.1.0 Release Video\n\n## Opening (0:00 - 0:30)\n\nNARRATOR:\nThe wrapper must avoid converting structural lin...e the final PDS create command with a highlighted\nhasVisual check and a callout for the normalized script artifact.\n'

tests/test_release_video.py:467: AssertionError
=========================== short test summary info ============================
FAILED tests/test_release_video.py::test_release_video_normalizes_label_only_visual_blocks
FAILED tests/test_release_video.py::test_release_video_does_not_swallow_unsafe_label_only_visual_blocks
2 failed, 1 warning in 0.39s
```

--- a/scripts/release_video.py
+++ b/scripts/release_video.py
@@ -2586,10 +2586,11 @@ def collapse_label_only_visual_cues(script: str) -> tuple[str, bool]:
     while index < len(lines):
         line = lines[index]
         if is_visual_label_only(line):
-            cue_index = next_nonblank_line_index(lines, index + 1)
-            if cue_index is not None and is_collapsible_visual_cue_line(lines[cue_index]):
-                append_spaced(normalized, f"VISUAL: {clean_visual_cue(lines[cue_index])}")
-                index = cue_index + 1
+            cue_lines, next_index = collect_visual_cue_paragraph(lines, index + 1)
+            if cue_lines:
+                cue = " ".join(cue_lines)
+                append_spaced(normalized, f"VISUAL: {cue}")
+                index = next_index
                 changed = True
                 continue
         normalized.append(line)
@@ -2600,12 +2601,26 @@ def collapse_label_only_visual_cues(script: str) -> tuple[str, bool]:
     return trim_repeated_blank_lines(normalized), True
 
 
-def next_nonblank_line_index(lines: list[str], start: int) -> int | None:
-    """Return the next nonblank line index at or after start."""
-    for index in range(start, len(lines)):
-        if lines[index].strip():
-            return index
-    return None
+def collect_visual_cue_paragraph(lines: list[str], start: int) -> tuple[list[str], int]:
+    """Collect cue paragraph lines after a label-only visual marker."""
+    cue_lines: list[str] = []
+    index = start
+    while index < len(lines) and not lines[index].strip():
+        index += 1
+    while index < len(lines):
+        line = lines[index]
+        if not line.strip():
+            break
+        if is_collapsible_visual_cue_line(line):
+            cue_lines.append(clean_visual_cue(line))
+            index += 1
+            continue
+        visual = visual_cue_text(line)
+        if visual and not cue_lines:
+            cue_lines.append(visual)
+            index += 1
+        break
+    return cue_lines, index
 
 
 def is_collapsible_visual_cue_line(line: str) -> bool:
@@ -2634,6 +2649,7 @@ def validate_release_video_script(
         "hasVisual": any(visual_cue_text(line) for line in script.splitlines()),
         "hasNoModelWrapperText": not has_unstripped_model_wrapper_text(script),
         "hasNoDuplicateNarratorLabels": not has_duplicate_narrator_labels(script),
+        "hasNoLabelOnlyVisualCues": not has_label_only_visual_cues(script),
         "hasNoMarkdownFences": not has_markdown_fence_line(script),
     }
     errors = [name for name, passed in checks.items() if not passed]
@@ -2662,6 +2678,11 @@ def has_duplicate_narrator_labels(script: str) -> bool:
         if line.strip():
             previous_pending_label = False
     return False
+
+
+def has_label_only_visual_cues(script: str) -> bool:
+    """Return whether the script still contains empty visual cue labels."""
+    return any(is_visual_label_only(line) for line in script.splitlines())
 
 
 def has_markdown_fence_line(script: str) -> bool:

--- a/tests/test_release_video.py
+++ b/tests/test_release_video.py
@@ -322,6 +322,10 @@ def test_release_video_generates_script_and_invokes_pds_publish(tmp_path: Path):
     assert "what changed, why it matters, and the practical business value" in claude_prompt
     assert "## Release hook (0:00 - 0:12)" in claude_prompt
     assert "Every spoken narration block starts with `NARRATOR:` on its own line." in claude_prompt
+    assert (
+        "Every non-spoken visual cue is written on one line as `VISUAL: <cue text>`."
+        in claude_prompt
+    )
     assert "release video automation" in claude_prompt
     claude_argv = json.loads((repo / "claude_argv.json").read_text(encoding="utf8"))
     assert claude_argv[claude_argv.index("--model") + 1] == "claude-opus-4-8"
@@ -411,6 +415,88 @@ Let me know if you want a punchier version.
     assert validation["checks"]["hasNoModelWrapperText"] is True
     assert validation["checks"]["hasNoDuplicateNarratorLabels"] is True
     assert validation["errors"] == []
+
+
+def test_release_video_normalizes_label_only_visual_blocks():
+    release_video = load_release_video_module()
+    script = """# PDD v1.1.0 Release Video
+
+## Opening
+
+NARRATOR:
+PDD v1.1.0 makes release-video recovery easier to audit by preserving the
+generated script, the normalized script, and the validation evidence before
+the PDS create step receives anything.
+
+VISUAL:
+
+show the release context, normalized script, and pds_run.json side by side with callout labels for raw output, final script, and validation state.
+
+## Recovery
+
+NARRATOR:
+Operators can compare the raw model output with the final script and quickly
+see which sanitation steps ran, which reduces guesswork during release
+recovery and keeps publishing diagnostics reproducible.
+
+VISUAL:
+zoom into release_video_script_validation.json highlighting checks.hasVisual true and the collapsed visual-label change entry.
+"""
+
+    artifacts = release_video.prepare_release_video_script(script, source="test")
+
+    assert "\nVISUAL:\n" not in artifacts["script"]
+    assert (
+        "\nVISUAL: show the release context, normalized script, and pds_run.json"
+        in artifacts["script"]
+    )
+    assert "\nVISUAL: zoom into release_video_script_validation.json" in artifacts["script"]
+    assert artifacts["validation"]["checks"]["hasVisual"] is True
+    assert "collapsed_label_only_visual_cues" in artifacts["validation"]["changes"]
+    assert artifacts["validation"]["errors"] == []
+
+
+def test_release_video_does_not_swallow_unsafe_label_only_visual_blocks():
+    release_video = load_release_video_module()
+    script = """# PDD v1.1.0 Release Video
+
+## Opening
+
+NARRATOR:
+The wrapper must avoid converting structural lines or wrapper prose into visual
+cues when a model emits an empty visual label without useful cue text after it.
+
+VISUAL:
+
+## Recovery
+
+VISUAL:
+
+NARRATOR:
+This narration remains a narrator block instead of becoming a visual cue,
+because the line after the empty visual label is another script label.
+
+VISUAL:
+
+Here is the release video script you asked for:
+
+VISUAL:
+
+show the validation JSON beside the final PDS create command with a highlighted
+hasVisual check and a callout for the normalized script artifact.
+"""
+
+    artifacts = release_video.prepare_release_video_script(script, source="test")
+
+    assert "\nVISUAL: ## Recovery" not in artifacts["script"]
+    assert "\nVISUAL: NARRATOR:" not in artifacts["script"]
+    assert "\nVISUAL: Here is the release video script" not in artifacts["script"]
+    assert (
+        "\nVISUAL: show the validation JSON beside the final PDS create command"
+        in artifacts["script"]
+    )
+    assert artifacts["validation"]["checks"]["hasVisual"] is True
+    assert artifacts["validation"]["checks"]["hasNoModelWrapperText"] is False
 
 
 def test_release_video_preserves_exact_raw_claude_output_artifact(tmp_path: Path):

--- a/tests/test_release_video.py
+++ b/tests/test_release_video.py
@@ -497,6 +497,34 @@ VISUAL: show release_video_script_validation.json with checks.hasVisual true.
     assert artifacts["validation"]["errors"] == []
 
 
+def test_release_video_normalizes_wrapped_same_line_visual_cue_continuation():
+    release_video = load_release_video_module()
+    script = """Hook: PDD v1.1.0 turns release-video recovery into an operator-visible
+path instead of a best-effort publish step that can fail without enough context.
+
+Narration: The release wrapper now keeps generated scripts, validation evidence,
+PDS run handles, status query diagnostics, and recovery commands together before
+the publish request reaches PDS. Maintainers can reattach to the same run, see
+whether a running sidecar is stale, and retry with a stable idempotency key
+without regenerating the release story or losing the incident trail.
+
+Visual direction: show the changelog, generated script, pds_run.json, status
+query output, and final YouTube receipt side by side.
+"""
+
+    artifacts = release_video.prepare_release_video_script(script, source="test")
+
+    assert (
+        "\nVISUAL: show the changelog, generated script, pds_run.json, status "
+        "query output, and final YouTube receipt side by side."
+        in artifacts["script"]
+    )
+    assert "\nNARRATOR:\nquery output" not in artifacts["script"]
+    assert "collapsed_wrapped_visual_cues" in artifacts["validation"]["changes"]
+    assert artifacts["validation"]["checks"]["hasVisual"] is True
+    assert artifacts["validation"]["errors"] == []
+
+
 def test_release_video_validation_rejects_leftover_label_only_visual_cues():
     release_video = load_release_video_module()
     script = """# PDD v1.1.0 Release Video

